ra/ra_test.go

```diff
@@ -844,8 +848,336 @@ func TestPerformValidationSuccess(t *testing.T) {
+}
+
+func (msa *mockSAPaused) GetPausedIdentifiers(ctx context.Context, req *sapb.RegistrationID, _ ...grpc.CallOption) (*sapb.Identifiers, error) {
+	msa.Lock()
+	defer msa.Unlock()
+	_, ok := msa.registrationsForRegID[req.Id]
+	if !ok {
+		return nil, fmt.Errorf("Unable to find registration for regID %d", req.Id)
+	}
+	idents, ok := msa.identifiersForRegID[req.Id]
+	if !ok {
+		return nil, fmt.Errorf("No identifiers paused yet")
+	}
+	return &sapb.Identifiers{Identifiers: idents}, nil
+}
+
+func (msa *mockSAPaused) FinalizeAuthorization2(ctx context.Context, req *sapb.FinalizeAuthorizationRequest, _ ...grpc.CallOption) (*emptypb.Empty, error) {
+	return &emptypb.Empty{}, nil
+}
+
+func TestPerformValidation_FailedValidationsTriggerPauseIdentifiersRatelimit(t *testing.T) {
+	if !strings.Contains(os.Getenv("BOULDER_CONFIG_DIR"), "test/config-next") {
+		t.Skip()
+	}
+
+	va, sa, ra, redisSrc, fc, cleanUp := initAuthorities(t)
+	defer cleanUp()
+
+	features.Set(features.Config{AutomaticallyPauseZombieClients: true})
+	defer features.Reset()
+
+	mockSA := newMockSAPaused(sa)
+	ra.SA = mockSA
+
+	// Override the default ratelimits to only allow one failed validation.
+	txnBuilder, err := ratelimits.NewTransactionBuilder("testdata/one-failed-validation-before-pausing.yml", "")
+	test.AssertNotError(t, err, "making transaction composer")
+	ra.txnBuilder = txnBuilder
+
+	// We know this is OK because of TestNewAuthorization
+	domain := "example.net"
+	authzPB := createPendingAuthorization(t, sa, domain, fc.Now().Add(12*time.Hour))
+	mockSA.registrationsForRegID[authzPB.RegistrationID] = Registration
+	mockSA.authorizationsForRegID[authzPB.RegistrationID] = authzPB
+
+	// We induce the bad path by setting a problem. This will consume all
+	// available capacity in the rate limit bucket.
+	va.PerformValidationRequestResultReturn = &vapb.ValidationResult{
+		Records: []*corepb.ValidationRecord{
+			{
+				AddressUsed:   []byte("192.168.0.1"),
+				Hostname:      domain,
+				Port:          "8080",
+				Url:           fmt.Sprintf("http://%s/", domain),
+				ResolverAddrs: []string{"rebound"},
+			},
+		},
+		Problems: &corepb.ProblemDetails{
+			Detail: fmt.Sprintf("CAA invalid for %s", domain),
+		},
+	}
+
+	challIdx := dnsChallIdx(t, authzPB.Challenges)
+	authzPB, err = ra.PerformValidation(ctx, &rapb.PerformValidationRequest{
+		Authz:          authzPB,
+		ChallengeIndex: challIdx,
+	})
+	test.AssertNotError(t, err, "PerformValidation failed")
+
+	var vaRequest *vapb.PerformValidationRequest
+	select {
+	case r := <-va.performValidationRequest:
+		vaRequest = r
+	case <-time.After(time.Second):
+		t.Fatal("Timed out waiting for DummyValidationAuthority.PerformValidation to complete")
+	}
+
+	// Verify that the VA got the request, and it's the same as the others
+	test.AssertEquals(t, authzPB.Challenges[challIdx].Type, vaRequest.Challenge.Type)
+	test.AssertEquals(t, authzPB.Challenges[challIdx].Token, vaRequest.Challenge.Token)
```

Comment:
I don't think we need to confirm this in the test.

Reply:
This is a copy/paste from `TestPerformValidationSuccess`. I can remove it.